portal-web/docroot/html/common/themes/top_head.jsp

```diff
@@ -27,16 +27,36 @@
 <link class="lfr-css-file" data-senna-track="temporary" href="<%= HtmlUtil.escapeAttribute(themeDisplay.getClayCSSURL()) %>" id="liferayAUICSS" rel="stylesheet" type="text/css" />
 
 <%
-List<Portlet> portlets = null;
+Set<Portlet> portlets = null;
 
 if (layoutTypePortlet != null) {
-	portlets = layoutTypePortlet.getAllPortlets();
+	portlets = new HashSet<>(layoutTypePortlet.getAllPortlets());
 }
 
 if (layout != null) {
 	String ppid = ParamUtil.getString(request, "p_p_id");
 
-	if ((layout.isTypeEmbedded() || layout.isTypePortlet()) && (themeDisplay.isStateMaximized() || themeDisplay.isStatePopUp() || (layout.isSystem() && Objects.equals(layout.getFriendlyURL(), PropsValues.CONTROL_PANEL_LAYOUT_FRIENDLY_URL)))) {
+	if((layout.isTypeAssetDisplay() || layout.isTypeContent())) {
+		List<com.liferay.portal.kernel.model.PortletPreferences> portletPreferencesList =
+			PortletPreferencesLocalServiceUtil.getPortletPreferences(
+				PortletKeys.PREFS_OWNER_ID_DEFAULT,
+				PortletKeys.PREFS_OWNER_TYPE_LAYOUT, layout.getPlid());
+
+		for (com.liferay.portal.kernel.model.PortletPreferences portletPreferences : portletPreferencesList) {
+			Portlet portlet = PortletLocalServiceUtil.getPortletById(
+				company.getCompanyId(),
+				portletPreferences.getPortletId());
+
+			if ((portlet == null) || !portlet.isActive() ||
+				portlet.isUndeployedPortlet()) {
+
+				continue;
+			}
+
+			portlets.add(portlet);
+		}
+	}
```

Comment:
Does this cover the portlets within fragments? I ask because there is a case where[ portlet paths are still written to the page](https://github.com/liferay/liferay-portal/blob/d1398e3f023046624df4488d897e39c065d073d4/modules/apps/fragment/fragment-impl/src/main/java/com/liferay/fragment/internal/renderer/FragmentEntryFragmentRenderer.java#L416-L423) from the fragment renderer, just wanted to be sure this is not overlapping with the other logic

Reply:
I just tested this with `<lfr-widget-rss></lfr-widget-rss>` in a fragment and it seems to work as expected

Reply:
@dsanz the display context that was removed is for view mode. The portlet paths that you mentioned and are still written are for edit mode. We would have to check if this is still required or not, since now we include them in the header